examples/BytePairEncoder.hs

```diff
@@ -0,0 +1,219 @@
+import Streamly.Internal.Data.Pipe (Pipe (..), Step (..))
+import Streamly.Internal.Data.Stream (pipe)
+import System.Environment (getArgs)
+
+-------------------------------------------------------------------------------
+-- Byte indexing and text representation
+-------------------------------------------------------------------------------
+
+-- Stores byte-sequence-to-index mapping and index-to-text mapping
+data ByteMappings = ByteMappings
+  { byteToIndex :: !(M.Map Word8 Int), -- Maps bytes to unique indices
+    seqToIndex :: !(M.Map (V.Vector Word8) Int), -- Maps sequences of bytes to unique indices
+    indexToText :: !(M.Map Int String), -- Maps indices to text representation
+    nextIndex :: !Int -- Next available index
+  }
+
+instance Show ByteMappings where
+  show (ByteMappings b2i _ i2t nidx) =
+    "ByteMappings:\n"
+      ++ "byteToIndex = "
+      ++ show b2i
+      ++ "\n"
+      ++ "indexToText = "
+      ++ show i2t
+      ++ "\n"
+      ++ "nextIndex = "
+      ++ show nidx
+
+{-# INLINE initializeSingleBytes #-}
+initializeSingleBytes :: (Monad m) => Stream m Word8 -> m ByteMappings
+initializeSingleBytes stream = do
+  -- Collect unique bytes and create initial mappings
+  uniqueBytes <-
+    stream
+      & Stream.fold (Fold.foldl' (\m b -> M.insert b () m) M.empty)
+
+  let bytes = V.fromList $ M.keys uniqueBytes
+      indices = [0 .. (V.length bytes - 1)]
+      b2i = M.fromList $ zip (V.toList bytes) indices
+      s2i = M.fromList $ zip (map V.singleton (V.toList bytes)) indices
+      i2t = M.fromList $ zip indices (map ((: []) . toEnum . fromIntegral) (V.toList bytes))
+
+  return $ ByteMappings b2i s2i i2t (length indices)
+
+{-# INLINE mapToIndexStream #-}
+mapToIndexStream :: (Monad m) => ByteMappings -> Stream m Word8 -> Stream m Int
+mapToIndexStream mapping = fmap (\k -> M.findWithDefault (-1) k (byteToIndex mapping))
+
+-------------------------------------------------------------------------------
+-- Count and merge most frequent pairs
+-------------------------------------------------------------------------------
+
+charToWord8 :: Char -> Word8
+charToWord8 = toEnum . fromEnum
+
+-- Stores pair frequencies for merging
+type PairFrequencies = M.Map (Int, Int) Int
+
+{-# INLINE countPairs #-}
+countPairs :: (MonadIO m) => Stream m Int -> m PairFrequencies
+countPairs stream =
+  stream
+    & Stream.chunksOf 2
+    & Stream.fold (Fold.foldl' addPair M.empty)
+  where
+    addPair acc chunk =
+      case Array.toList chunk of
+        [b1, b2] -> M.insertWith (+) (b1, b2) 1 acc
+        _ -> acc
+
+mostFrequentPair :: PairFrequencies -> ((Int, Int), Int)
+mostFrequentPair = maximumBy (comparing snd) . M.toList
+
+updateMappings :: ByteMappings -> (Int, Int) -> ByteMappings
+updateMappings (ByteMappings b2i s2i i2t nidx) (i1, i2) =
+  let text1 = M.findWithDefault "?" i1 i2t
+      text2 = M.findWithDefault "?" i2 i2t
+      newToken = text1 ++ text2
+      bytes = V.fromList $ map charToWord8 newToken
```

Comment:
This looks incorrect. `Char` is essentially 4 bytes. You are losing information here.
```
Char -> [Word8]
```